### PR TITLE
CORTX-30806: Feature: Added version compatibility check

### PR DESCRIPTION
**Pull request by kailashyadav11**
Signed-off-by: Kailash Yadav <kailash.yadav@seagate.com>

# Problem Statement
-  Implement Version compatibility logic.
- For example, current version: 10, new version: 12
  if version 12 config allow transition to and from 10 and 11 version then version 10 is said to be compatible with version 12

# Design
-  For Bug describe the fix here. 
-  For Feature, Post the link to the solution page on the confluence CORTX Foundation Library 
-  https://jts.seagate.com/browse/CORTX-30806

# Coding 
-  Coding conventions are followed and code is consistent [Y/N]: Y
-  Confirm All CODACY errors are resolved [Y/N]: Y

# Testing
- Are test cases updated along with code changes due to Enhancements/Bugs [Y/N]:
- Confirm that new test cases are added to regression and sanity plan files and relevant feature plan files [Y/N]:
- Confirm that Test Cases are added for new features added [Y/N]: 
- Confirm Test Cases cover Happy Path, Non-Happy Path and Scalability [Y/N]: 
- Confirm Testing was performed with installed RPM/K8s deployment [Y/N]: 

# Review Checklist 
####  Before posting the PR please ensure:
- [ ] PR is self reviewed
- [ ] Is there a change in filename/package/module or signature [Y/N]: 
- [ ] If yes for above point, Is a notification sent to all other cortx components [Y/N]
- [ ] New package/s added to setup.py?
- [ ] Jira is updated
- [ ] Check if the description is clear and explained. 
- [ ] Check Acceptance Criterion is defined. 
- [ ] All the tests performed should be mentioned before Resolving a JIRA. 
- [ ] Verification needs to be done before marked as Closed/Verified.

# KVstore/Confstore feature (changes/fixes) checklist
#### Confirm changes are made for:
- [ ] ConfStore
- [ ] KVStore
- [ ] ConfCli
- [ ] Test cases added for all above 3
- [ ] changes done in all the KVpayloads (ConsulKvPayload, IniKvPayload, KvPayload)

# Documentation
- [ ] Changes done to WIKI / Confluence page

### CODE REVIEW

File: py-utils/src/utils/schema/release.py [hunk 1]
```diff
@@ -91,6 +91,24 @@ def version_check(self, deploy_version: str, release_version: str):
                 break
         return ret_code
 
+    def is_version_compatible(self, deploy_version:str, component:str):
```
Comment (nitin-seagate) [inline py-utils/src/utils/schema/release.py:94]: Lets rename function: `is_update_version_compatible`
Which accepts arguments: deploy_version, update_version and component as parameters...
In essence we compare deploy_version and update_version

**Review by nitin-seagate — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/schema/release.py (hunk 1 above)
Comment (kailashyadav11) [reply py-utils/src/utils/schema/release.py:94]: @nitin-seagate In this API we do not need update_version for compatibilty instead we are using RELEASE spec of the update image.

Release Metadata contains compatibility information. User should be able to query using CLI and validate upgrade compatibility with the installed version
Release Metadata will be available in RELEASE.INFO in Key-Value format (to be read using conf store)
REQUIRES:
    - CORTX >= 2.0.0
    - cortx-motr >= 2.0.0-123
    - cortx-s3server >= 2.0.0-123
    - cortx-hare >= 2.0.0-123
    - cortx-py-utils >= 2.0.0-123
    - cortx-csm_agent >= 2.0.0-123
    - cortx-csm_web >= 2.0.0-123
    - cortx-sspl >= 2.0.0-123
    - cortx-ha >= 2.0.0-123
    - cortx-prvsnr >= 2.0.0-123

**Review by nitin-seagate — APPROVED** [review]
LGTM

**Review by tanujashinde0405 — APPROVED** [review]

File: py-utils/src/utils/schema/release.py [hunk 2]
```diff
@@ -91,6 +91,25 @@ def version_check(self, deploy_version: str, release_version: str):
...
+
+        Returns True if version is compatible for upgrade/downgrade else returns False.
+        Parameters:
+        upgrade_version - Deployed image version.
+        component - Component to be checked for version compatibility.
+        """
+        upgrade_version = self._get_val('VERSION')
+        min_compatible_versions = self._get_val('REQUIRES')
+        try:
+            min_compatible_version = [ x.split('>=')[1].strip() for x in min_compatible_versions if component in x][0]
```
Comment (selvakumaar5496) [inline py-utils/src/utils/schema/release.py:106]: trailing space can be removed 
[ x.split('>=')[1].strip() for x in min_compatible_versions if component in x][0] ==> [x.split('>=')[1].strip() for x in min_compatible_versions if component in x][0]

**Review by selvakumaar5496 — APPROVED** [review]

**Review by ujjwalpl — APPROVED** [review]

**Pull request merged**